이메일 인증 리팩토링

kihyuk-jeong opened the pull request:
![인증메일재전송오븐뷰](https://user-images.githubusercontent.com/39195377/108088275-9f600f00-70bb-11eb-9a30-51419c0f1bb5.PNG)

1. 회원가입시 랜덤으로 생성된 토큰 URL을 이메일로 전송
2. 인증하지 않은 회원은 판매/구매 기능 제한 (추후 추가될 판매/구매 관련 핸들러에 `@EmailAuthCheck`)
3. 메모리 낭비를 방지하기 위해 Redis에 저장된 토큰의 timeout을 24시간으로 설정
4. 24시간 내에 인증을 못하였을 경우, 이메일 인증 토큰을 재전송 가능하도록 수정


<리뷰 반영 내용>
1. 불필요한 부정 연산자 제거
2. 외부 주입 방법 변경
3. 호스트명 하드코딩 -> 상수 처리
4. 인증 토큰 만료시간 설정 및 비즈니스 로직 수정
5. 불필요한 공백 제거
6. 코딩 컨벤션 수정

kihyuk-jeong committed:
[#22] feat : 이메일 인증 및 인증 관련 상수 클래스로 관리

1. 회원가입시 등록된 이메일로 랜덤 토큰 발송
2. 토큰 일치 여부 확인 후 인증처리
3. 인증 관련 상수 클래스로 관리

 #22 : 이메일 인증, 인증 관련 상수 클래스로 관리

kihyuk-jeong committed:
[#26] refactor : 리뷰 반영 / 비즈니스 로직 수정

1. 불필요한 부정 연산자 제거
2. 외부 주입 방법 변경
3. 호스트명 하드코딩 -> 상수 처리
4. 인증 토큰 만료시간 설정 및 비즈니스 로직 수정
5. 불필요한 공백 제거
6. 코딩 컨벤션 수정

 #26 : 이메일 인증 리팩토링

kihyuk-jeong committed:
[#26] refactor : 불필요한 클래스 파일 삭제

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/annotation/EmailAuthCheck.java:
```diff
@@ -0,0 +1,19 @@
+
+@Retention(RUNTIME)
+@Target(METHOD)
+public @interface EmailAuthCheck {
```
이걸 왜 AOP로 빼야하는지 모르겠습니다~ 이메일 인증을 완료한 사람인지 여부를 매 번 체크한다면 의미없는 부하를 가하는게 되지 않을까요? 로그인시에 한번만 체크하게 하는게 좋을 것 같습니다~

kihyuk-jeong replied:
생각해보니 이메일 인증 여부를 매 번 체크할 필요가 없는 것 같습니다. 
기존에 사용하던 `LoginCheck` 에 열거형 타입을 추가해 이메일 인증이 필수로 요구되는 핸들러에만 로그인 인터셉터에서 추가적으로 체크하도록 수정하겠습니다.

f-lab-dev replied:
LoginCheck에서 검사하더라도 매번 검사하는건 바뀌지 않습니다. 로그인 로직에 껴넣고 인증이 완료되지 않았으면 로그인을 실패하게 만드는게 낫지 않을까요?

kihyuk-jeong replied:
@Hosick 님과 의논해봤는데 이메일 인증을 완료하지 않더라도 커뮤니티성으로 서비스를 이용하도록 하고 주요 기능인 판매/구매 기능만 제한하는 방향으로 구현하려고 합니다.  

만약 이런식의 로직이면 로그인을 체크하는 Interceptor에서 인증 여부를 한번 더 확인하는것이 판매/구매 관련 리소스마다 인증여부를 확인하는 것 보다 낫다고 판단해서 이렇게 구현했습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/config/AppProperties.java:
```diff
@@ -0,0 +1,19 @@
+@Getter
+@RequiredArgsConstructor
+@ConstructorBinding
+@ConfigurationProperties("certification-related-constants")
```
property는 외부에서 주입해주는건데 `constants`라는 단어를 붙여주는게 맞을까요? 그리고 `related`라는 단어도 뜻이 애매해보입니다. `조금이라도 연관된게 있으면 다 여기에 넣겠다`라는 뜻으로도 받아들여질 수 있을 것 같아서요~

kihyuk-jeong replied:
음 인증과 관련된 값들을 외부에서 주입받는다는 의미로 `external-certification`  으로 설정하면 괜찮을까요?? 네이밍 부분에서 고민을 많이 하는데도 좋은 이름이 잘 생각나지 않네요

f-lab-dev replied:
외부에서 주입하고말고는 개발자의 마음입니다~ 그런것은 드러나지 않아도 됩니다.
정확한 역할을 나타내려면 어떻게 해야할지 고민해보시면 좋을 것 같아요~

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/LoginService.java:
```diff
@@ -28,7 +28,9 @@ public void existByEmailAndPassword(LoginRequest loginRequest) {
     }
 
-    public void login(String email) {
+    public void login(LoginRequest loginRequest) {
```
이 클래스는 세션을 이용한 로그인 서비스이므로 `SessionLoginService`로 이름짓는게 낫지 않을까요?

kihyuk-jeong replied:
네 조금 더 명확하게 짓는 게 좋을 것 같습니다. 리뷰 반영하겠습니다.

kihyuk-jeong committed:
[#26] refactor : 리뷰 반영

1. LoginCheckInterceptor에서 authority가 AUTH인 경우 이메일 인증 여부를 추가적으로 확인하도록 수정
2. property 외부 주입 이름 수정
3. loginService -> sessionLoginService

 #26 : 이메일 인증 리팩토링

f-lab-dev reviewed (commented):
위 코멘트에서 적었듯이 로그인 여부 검사가 아니라 로그인할때 검사하도록 해주시는게 좋을 것 같네요~

kihyuk-jeong committed:
chore : 공백 제거, 컨벤션 수정

kihyuk-jeong committed:
[#26] refactor : 이메일 인증 리뷰 반영

1. 로그인 시 이메일 인증 여부를 session에 함께 저장
2. 유저 권한을 인증/미인증으로 분류하여 미인증시 특정 서비스 접근 불가능하도록 구현

 #26 : 이메일 인증 리팩토링

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/annotation/LoginCheck.java:
```diff
@@ -17,8 +17,8 @@
 
     enum EmailAuthStatus {
-        FREE, AUTH
+        UN_AUTH, AUTH
```
이거는 `_`로 구분해주지 않아도 될 것 같아요~

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/SessionLoginService.java:
```diff
@@ -32,18 +34,31 @@ public void login(LoginRequest loginRequest) {
         return (String) session.getAttribute(USER_ID);
     }
 
+    public boolean getEmailAuth() {
```
boolean을 리턴할때는 앞에 `is`를 붙여주는게 좋습니다~

kihyuk-jeong committed:
[#26] refactor : 리뷰 반영
